index.js

```diff
@@ -1,483 +1,668 @@
   }
 
-  _ontick () {
-    this._tick++
-    if ((this._tick & 7) === 0) this._pingSome()
-    if ((this._tick & 63) === 0 && this.nodes.length < 20) this.bootstrap()
+  onmessage (buf, rinfo) {
+    if (buf.byteLength > 1) this.rpc.onmessage(buf, rinfo)
   }
 
-  address () {
-    return this.socket.address()
+  sampledNAT () {
+    return this._sampled
   }
 
-  command (name, opts) {
-    this._commands.set(name, {
-      inputEncoding: codecs(opts.inputEncoding || opts.valueEncoding),
-      outputEncoding: codecs(opts.outputEncoding || opts.valueEncoding),
-      query: opts.query || queryNotSupported,
-      update: opts.update || updateNotSupported
-    })
+  ready () {
+    return this._bootstrapping
   }
 
-  ready (onready) {
-    if (!this.bootstrapped) this.once('ready', onready)
-    else onready()
+  query (target, command, value, opts) {
+    this._refreshTicks = REFRESH_TICKS
+    return new Query(this, target, command, value || null, opts)
   }
 
-  onrequest (type, message, peer) {
-    if (validateId(message.id)) {
-      this._addNode(message.id, peer, null, message.to)
-    }
+  ping (node) {
+    return this.request(null, 'ping', null, node)
+  }
 
-    switch (message.command) {
-      case '_ping':
-        return this._onping(message, peer)
+  request (target, command, value, to, opts) {
+    const ephemeral = this.ephemeral || !!(opts && opts.socket !== this.rpc.socket)
+    const token = (opts && opts.token) || null
 
-      case '_find_node':
-        return this._onfindnode(message, peer)
+    return this.rpc.request({
+      version: 2,
+      tid: 0,
+      from: null,
+      to,
+      id: ephemeral ? null : this.table.id,
+      token,
+      target,
+      closerNodes: null,
+      command,
+      status: 0,
+      value
+    }, opts)
+  }
 
-      case '_holepunch':
-        return this._onholepunch(message, peer)
+  requestAll (target, command, value, nodes, opts = {}) {
+    const min = typeof opts.min === 'number' ? opts.min : 1
+    if (nodes.length < min) return Promise.reject(new Error('Too few nodes to request'))
 
-      default:
-        return this._oncommand(type, message, peer)
-    }
+    const p = []
+    for (const node of nodes) p.push(this.request(target, command, value, node))
+    return race(p, min, opts.max)
   }
 
-  _onping (message, peer) {
-    if (message.value && !this.id.equals(message.value)) return
-    this._io.response(message, peers.encode([peer]), null, peer)
+  destroy () {
+    this.rpc.destroy()
+    if (this._resolveSampled !== null) {
+      this._resolveSampled(false)
+      this._resolveSampled = null
+    }
+    if (this._localSocket) {
+      this._localSocket.close()
+      this._localSocket = null
+    }
+    clearInterval(this._tickInterval)
   }
 
-  _onholepunch (message, peer) {
-    const value = decodeHolepunch(message.value)
-    if (!value) return
+  async bootstrap () {
+    await Promise.resolve() // wait a tick, so apis can be used from the outside
```

Comment:
is there potential for this to fail in scenarios where a bootstrap is done after awaiting some other promise in the user code?